### PR TITLE
docs: Add RT-DETR variant to per-layer learning rates guide

### PR DESCRIPTION
**Pull request by artest08**
Adds an RT-DETR variant of the per-layer learning rate example in the custom trainer guide. The new `RTDETRBackboneLRTrainer` subclass reads backbone length from `model.yaml["backbone"]` and routes parameters into per-section, per-kind groups so the backbone trains at `lr * backbone_lr_ratio` while the head trains at `lr`, a common recipe when fine-tuning a pretrained backbone alongside a freshly-initialized decoder.

```python
from ultralytics import RTDETR

# RTDETRBackboneLRTrainer defined in the docs guide
RTDETR("rtdetr-l.pt").train(data="coco8.yaml", epochs=20, trainer=RTDETRBackboneLRTrainer)
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
📘 This PR expands the custom trainer guide with a new RT-DETR example showing how to fine-tune backbone and head layers with different learning rates, while also adding a new docs author entry.

### 📊 Key Changes
- Added a new **RT-DETR Variant** section to the [custom trainer guide](https://docs.ultralytics.com/guides/custom-trainer/) 🧠
- Introduced a full example of a custom `RTDETRTrainer` subclass that:
  - applies a **lower learning rate to backbone layers**
  - automatically detects backbone depth from `model.yaml["backbone"]`
  - works across different RT-DETR variants without hardcoded layer counts
- Split optimizer parameters into **six groups**: head/backbone × weight/BatchNorm/bias ⚙️
- Matched the default training policy by:
  - disabling weight decay for BatchNorm parameters and biases
  - using RT-DETR-specific optimizer behavior such as `AdamW` defaults and bias warmup from `0.0`
- Added guidance on choosing `backbone_lr_ratio`, with practical recommendations for small vs. large pretrained backbones 💡
- Added a new contributor entry in `docs/mkdocs_github_authors.yaml` for @AffanBinFaisal 👤

### 🎯 Purpose & Impact
- Helps users fine-tune RT-DETR models more effectively by protecting pretrained backbone features while training newer head layers faster 🚀
- Makes the example more reusable across RT-DETR model variants, reducing manual edits and hardcoded assumptions
- Improves training best practices in the documentation, especially for advanced users customizing optimizers and learning-rate behavior
- Lowers the barrier for developers who want to build custom RT-DETR trainers by providing a ready-to-adapt template 🛠️
- Has no direct model-code behavior change; the main impact is **better documentation and clearer guidance** for training customization 📚

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist you soon. Please review the checklist below to help ensure a smooth integration 😊

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: docs/en/guides/custom-trainer.md [hunk 1]
```diff
@@ -314,6 +314,90 @@ model = YOLO("yolo26n.pt")
 model.train(data="coco8.yaml", epochs=20, trainer=PerLayerLRTrainer)
 ```
 
+### RT-DETR Variant
+
+For RT-DETR the pattern is the same with two refinements. The backbone length is read from `model.yaml["backbone"]` so the same trainer works across RT-DETR variants (RT-DETR-L, RT-DETR-X, ResNet-50/101 backbones) without hardcoding layer counts. Parameters are also split into weight, BatchNorm, and bias groups within each section so weight decay applies only to convolutional weights, matching the default trainer's policy. This is especially useful for RT-DETR fine-tuning, where the decoder head is typically randomly initialized while the backbone carries pretrained features that benefit from a lower learning rate:
```
Comment (UltralyticsAssistant) [inline docs/en/guides/custom-trainer.md]: 📝 **LOW**: The prose says weight decay applies "only to convolutional weights", but the code actually puts every non-norm, non-bias parameter into the `weight` groups. That includes Linear weights in the RT-DETR encoder/decoder, so this description overstates what the example is doing.

File: docs/en/guides/custom-trainer.md [hunk 2]
```diff
@@ -314,6 +314,90 @@ model = YOLO("yolo26n.pt")
...
+            raise NotImplementedError(f"This trainer only supports AdamW-family optimizers; got {name}")
+
+        # Identify backbone parameters from model.yaml and route each param into a (section, kind) group
+        unwrapped = unwrap_model(model)
+        backbone_len = len(unwrapped.yaml["backbone"])
+        norm_types = tuple(v for k, v in nn.__dict__.items() if "Norm" in k)
+        groups = {f"{s}_{k}": [] for s in ("head", "backbone") for k in ("weight", "bn", "bias")}
+
+        for module_name, module in unwrapped.named_modules():
+            for param_name, param in module.named_parameters(recurse=False):
```
Comment (UltralyticsAssistant) [inline docs/en/guides/custom-trainer.md:356]: 📝 **LOW**: Unlike the generic per-layer LR example above, this loop never skips `requires_grad=False` parameters. If users combine this trainer with `freeze=` or manual freezing, frozen tensors still get added to the optimizer groups and the logged group counts stop reflecting the actually trainable params. Adding the same guard here would keep the example aligned with trainable parameters only.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice addition overall—the dynamic backbone split is a useful RT-DETR-specific example, and the grouping logic is easy to follow. I only found a few docs/code alignment issues: two wording inaccuracies in the explanatory text and one small mismatch with the earlier example around filtering frozen parameters.

💬 Posted 2 inline comments

File: docs/en/guides/custom-trainer.md (hunk 1 above)
Comment (artest08) [reply docs/en/guides/custom-trainer.md]: This one is corrected!

File: docs/en/guides/custom-trainer.md (hunk 2 above)
Comment (artest08) [reply docs/en/guides/custom-trainer.md:356]: This is also integrated following the suggestions of assistant!

**Review by Laughing-q — APPROVED** [review]

**Comment by raimbekovm** [comment]
should the new optimizer param groups also carry the `"param_group": "weight" | "bn" | "bias"` tag that `BaseTrainer.build_optimizer` sets at [trainer.py:1031-1033](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/trainer.py#L1031-L1033)?

the warmup loop at [trainer.py:423](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/trainer.py#L423) keys off it to start bias groups from `warmup_bias_lr` and ramp them down to `lr`, which is the behavior described in the `# Bias lr falls from 0.1 to lr0, all other lrs rise from 0.0 to lr0` comment a few lines above. without the tag, all 6 of the new groups (including the 226 bias params on rtdetr-l) warm up from `0.0` like weights instead.

the `name="auto"` branch hides it (it forces `warmup_bias_lr = 0.0`), but i think it triggers when a user picks AdamW explicitly:

```python
from ultralytics import RTDETR
RTDETR("rtdetr-l.pt").train(
    data="coco8.yaml", epochs=20, trainer=RTDETRBackboneLRTrainer,
    optimizer="AdamW",  # skips the auto branch, keeps default warmup_bias_lr=0.1
)
```

replaying the warmup loop on rtdetr-l (`warmup_bias_lr=0.1`, `nw=500`, `lr=1e-4`):

| group | ni=0 | ni=250 | ni=500 |
|---|---|---|---|
| pr bias groups (226 params)     | 0.00000 | 0.00005 | 0.00010 |
| default bias group (226 params) | 0.10000 | 0.05005 | 0.00010 |

would it make sense to tag the bias groups (and the weight/bn groups for symmetry) so warmup behaves the same as the default trainer here, or is leaving them untagged intentional because this trainer is meant to lock the AdamW-only path?

**Comment by artest08** [comment]
@raimbekovm 

Let me restate your concern to make sure I follow: without the param_group tags, BaseTrainer's warmup loop can't identify which groups are biases, so all six groups (including bias) ramp up from 0.0 instead of starting from warmup_bias_lr. Is that right?

If so, I've addressed it two ways:

1. Added the "param_group" tags to all six group dicts so the warmup loop can route them correctly.
2. Moved self.args.warmup_bias_lr = 0.0 out of the auto branch so it applies on every optimizer path, since RT-DETR's recipe warms biases from 0 regardless of the optimizer chosen (assuming the optimizer is always ADAMW or similar type).

Does this address your concern?

**Comment by raimbekovm** [comment]
@artest08 yes, exactly that. just verified the new code on rtdetr-l: 6/6 groups carry `param_group`, 2 tagged `bias`, and with `warmup_bias_lr=0.0` unconditional all groups ramp linearly 0 -> initial_lr like the RT-DETR recipe expects. counter-check confirmed the tags still work if someone overrides `warmup_bias_lr` via cfg later (bias groups would correctly descend from the override down to lr0). both changes are defensive in a good way. lgtm from my side.

**Review by raimbekovm — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Merged with enthusiasm — thank you @artest08, @Laughing-q, and @raimbekovm! 🎉

As Benjamin Franklin said, “An investment in knowledge pays the best interest.” This PR is a great example of that: the new RT-DETR custom trainer guide makes advanced fine-tuning strategies much more accessible, practical, and reusable for the community.

Really appreciate the thoughtful documentation improvements, the robust optimizer grouping example, and the added guidance that helps users adapt RT-DETR training across variants with confidence. Also, thanks for adding the new docs author entry for @AffanBinFaisal.

Excellent contribution to the docs and developer experience. 📘🚀
